File: 121. Best Time to Buy and Sell Stock/Python/Best Time to Buy and Sell Stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        max_ending_here = 0
        max_so_far = 0
        for i in range(1, len(prices)):
            max_ending_here = max(0, max_ending_here + (prices[i] - prices[i - 1]))
            max_so_far = max(max_so_far, max_ending_here)
        return max_so_far

File: 121. Best Time to Buy and Sell Stock/C++/Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int maxEndingHere = 0;
        int maxSoFar = 0;
        for (int i = 1; i < prices.size(); i++) {
            maxEndingHere = max(0, maxEndingHere + prices[i] - prices[i - 1]);
            maxSoFar = max(maxSoFar, maxEndingHere);
        }
        return maxSoFar;
    }
};